alibuild_helpers/build.py

```diff
@@ -344,33 +344,29 @@ def doBuild(args, parser):
                   "  git pull --rebase\n",
                   pwd=os.getcwd(), star=star()))
 
-  # Prefetch git heads in parallel
-  debug("Prefetching git heads in parallel:")
-
-  fetchers = list(filter(lambda p: "source" in specs[p], buildOrder))
-  poolConfig = {"processes": min(max(len(fetchers), 1), 16)}
-  pool = ThreadPool(**poolConfig)
-  debug("Created thread pool with config: %s" % poolConfig)
+  # Clone/update repos
+  reposWithSource = list(filter(lambda p: "source" in specs[p], buildOrder))
+  if args.offline:
+    banner("Offline mode - not attempting to update any repositories\n")
+    for p in reposWithSource:
+      specs[p]["reference"] = "%s/%s" % (abspath(args.referenceSources), p.lower())
+      dieOnError(not exists(specs[p]["reference"]), "Local clone of %s does not exist yet, run again without --offline option." % specs[p]["package"]);
+  else:
+    for p in reposWithSource:
+      updateReferenceRepos(args.referenceSources, p, specs[p])
 
-  git_commands = {} # store futures
-  for p in fetchers:
-    # Replace source with local one if we are in development mode.
+  # Retrieve git heads
+  git_head_fetchers = list(filter(lambda p: "reference" in specs[p], buildOrder))
+  for p in git_head_fetchers:
+    cmd = "git ls-remote --heads %s" % specs[p]["reference"]
     if specs[p]["package"] in develPkgs:
       specs[p]["source"] = join(os.getcwd(), specs[p]["package"])
-
-    cmd = "git ls-remote --heads %s" % specs[p]["source"]
-    debug("Enqueued '%s'" % cmd)
-    git_commands[p] = pool.apply_async(getstatusoutput, (cmd,))
-
-  results = {}
-  for p, future in git_commands.items():
-    results[p] = future.get()
-
-  for p, result in results.items():
-    dieOnError(result[0], "Unable to fetch from %s" % specs[p]["source"])
-    specs[p]["git_heads"] = result[1].split("\n")
-
-  debug("Finished prefetching git heads")
+      cmd = "git ls-remote --heads %s" % specs[p]["source"]
+    res, output = getstatusoutput(cmd)
+    if sys.version_info < (3, 0):
```

Comment:
We have a `to_unicode` function in utilities, and a `getStatusOutputBash` that should be used instead (no need to check for Python versions here nor to `decode`).